ws/client.go

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package ws
+
+import (
+	"github.com/gorilla/websocket"
+	"github.com/mainflux/mainflux/pkg/messaging"
+)
+
+// Client handles messaging and websocket connection
+type Client struct {
+	conn *websocket.Conn
+	id   string
+}
+
+// NewClient returns a new Client object
+func NewClient(c *websocket.Conn) *Client {
+	return &Client{
+		conn: c,
+		id:   "",
+	}
+}
+
+// Cancel handles the websocket connection after unsubscribing
+func (c *Client) Cancel() error {
+	if c.conn == nil {
+		return nil
+	}
+	return c.conn.Close()
+}
+
+// Handle handles the sending and receiving of messages via the broker
+func (c *Client) Handle(msg messaging.Message) error {
+	// To prevent publisher from receiving its own published message
+	if msg.GetPublisher() == c.id {
+		return nil
+	}
+
+	return c.conn.WriteMessage(websocket.TextMessage, msg.Payload)
+}
```

Comment:
This is not covered by tests.

Reply:
I can create a new test `TestPubSub` with a channel, that checks if message is received but that will be kind of repitition of the `pkg/messaging/pubsub` tests.
Should I add this test?

Reply:
I meant more like: check if `p.conn.WriteMessage` works correctly - focus on WebSockets here. Like: creating a connection, waiting for messages (plain WebsSokets), and checking if Handle will send you the message.

Reply:
I see, but since this looks like an integration test, should I create a new test file for this, or just add another function to ws/api/endpoint_test.go ?

I have a small ws-client program, I'll modify that code to test the handle function

Reply:
I have to still add testcases list, and make some changes, but this is the approach I'm taking ->

Inside adapter_test.go, I've added a new function, to test the handler ->

If svc.Publish works -> connection.ReadMessage works -> c.conn.WriteMesssage inside `handle()` also works.
And I'll add test cases where we publish to ourselves (same chanID, thingKey, topic, etc.) To check if the extra condition inside handle also works.

Wanted to check if this is the right approach before progressing further

```
msgChan = make(chan []byte)
done    = make(chan interface{})

...
(Other vars and constants)
(TestPublish, TestSubscribe, TestUnsubscribe)
...

func TestWebsocketConn(t *testing.T) {
	thingsClient := httpmock.NewThingsClient(map[string]string{thingKey: chanID})
	svc, _ := newService(thingsClient)

	conn, _, err := websocket.DefaultDialer.Dial(socketUrl, nil)
	// defer conn.Close()
	assert.Nil(t, err, fmt.Sprintln("expected nil error when creating new ws connection"))
	c := ws.NewClient(conn)

	err = svc.Subscribe(context.Background(), thingKey, chanID, "", c)
	assert.Nil(t, err, fmt.Sprintln("expected nil error when creating new ws connection"))
	err = svc.Subscribe(context.Background(), thingKey, chanID, msg.Subtopic, c)
	assert.Nil(t, err, fmt.Sprintln("expected nil error when creating new ws connection"))

	go receiveHandler(t, conn)

	err = svc.Publish(context.Background(), thingKey, msg)
	assert.Nil(t, err, fmt.Sprintln("expected nil error when publishing"))
	data := <-msgChan
	receivedMsg := messaging.Message{}
	err = json.Unmarshal(data, &receivedMsg)
	assert.Nil(t, err, fmt.Sprintf("expected no error while unmarshalling: %s", err))
	assert.Equal(t, msg, receivedMsg, fmt.Sprintf("expected %+v, got %+v", msg, receivedMsg))

}

func receiveHandler(t *testing.T, connection *websocket.Conn) {
	defer close(done)

	for {
		_, msg, err := connection.ReadMessage()
		assert.Nil(t, err, fmt.Sprintf("unexpected error while reading ws message: %s", err))
		msgChan <- msg
	}
}

```